xdsl/dialects/riscv.py

```diff
@@ -215,6 +215,137 @@ class RISCVOp(Operation, ABC):
     def assembly_line(self) -> str | None:
         raise NotImplementedError()
 
+    def print_attributes(self, printer: Printer):
+        attributes = self.attributes
+        if (csr := attributes.get("csr")) is not None:
```

Comment:
This seems like an inverted logic using inheritance in a somewhat lopsided way.
I'd prefer to implement the printing of attributes for each class that owns/defines them.
This avoids all these existence checks and the printing is close to the code that it prints.

So, for example, I'd implement the csr stuff in the `CsrReadWriteOperation` class.

Reply:
I agree, instead of one huge print method we should have one basic RISC-V one (that falls back to printing an attr dict) and then specific ones for each specific format (csr, jumps, etc).